### PR TITLE
Adapt remove_copy(_if) to C++20

### PR DESCRIPTION
**Pull request by gonidelis**
This adapts `remove_copy` and `remove_copy_if` to C++20 by providing the ranges overloads according to the Standard. 

Note: _It is based on #5125 in order for the iterator-sentinel_value test cases to run (due to iter_sent.hpp changes) and thus this particular test is commented out for the time being. When #5125 will be merged, proper rebasing will take place._

Proper documentation will be added soon.

### CODE REVIEW

File: libs/parallelism/algorithms/include/hpx/parallel/algorithms/remove_copy.hpp [hunk 1]
```diff
@@ -35,10 +35,10 @@ namespace hpx { namespace parallel { inline namespace v1 {
         /// \cond NOINTERNAL
 
         // sequential remove_copy
-        template <typename InIter, typename OutIter, typename T, typename Proj>
+        template <typename InIter, typename Sent, typename OutIter, typename T,
+            typename Proj>
         inline util::in_out_result<InIter, OutIter> sequential_remove_copy(
```
Comment (hkaiser) [inline libs/parallelism/algorithms/include/hpx/parallel/algorithms/remove_copy.hpp:40]: This algorithm can be marked as `constexpr`.

File: libs/parallelism/algorithms/include/hpx/parallel/algorithms/remove_copy.hpp [hunk 2]
```diff
@@ -343,3 +344,111 @@ namespace hpx { namespace parallel { inline namespace v1 {
...
+    ///////////////////////////////////////////////////////////////////////////
+    // CPO for hpx::remove_copy_if
+    HPX_INLINE_CONSTEXPR_VARIABLE struct remove_copy_if_t final
+      : hpx::functional::tag<remove_copy_if_t>
+    {
+        // clang-format off
+        template <typename InIter, typename OutIter,
+            typename Pred, HPX_CONCEPT_REQUIRES_(
+                hpx::traits::is_iterator<InIter>::value &&
+                hpx::traits::is_iterator<OutIter>::value
```
Comment (hkaiser) [inline libs/parallelism/algorithms/include/hpx/parallel/algorithms/remove_copy.hpp:358]: Please add a check for the predicate here.

File: libs/parallelism/algorithms/include/hpx/parallel/algorithms/remove_copy.hpp [hunk 3]
```diff
@@ -343,3 +344,111 @@ namespace hpx { namespace parallel { inline namespace v1 {
...
+        template <typename InIter, typename OutIter,
+            typename T, HPX_CONCEPT_REQUIRES_(
+                hpx::traits::is_iterator<InIter>::value &&
+                hpx::traits::is_iterator<OutIter>::value
+            )>
+        // clang-format on
+        friend OutIter tag_invoke(hpx::remove_copy_t, InIter first, InIter last,
+            OutIter dest, T const& value)
+        {
+            typedef typename std::iterator_traits<InIter>::value_type Type;
```
Comment (hkaiser) [inline libs/parallelism/algorithms/include/hpx/parallel/algorithms/remove_copy.hpp:427]: Could you please add `static_assert`s verifying the validity of the iterators?

File: libs/parallelism/algorithms/tests/unit/container_algorithms/remove_copy_if_range.cpp [hunk 4]
```diff
@@ -16,9 +16,70 @@
...
+
+//     auto pred = [](std::int16_t const& a) -> bool { return a % 42 == 0; };
+
+//     auto pre_result = std::count_if(std::begin(c), std::end(c), pred);
+//     hpx::ranges::remove_copy_if(
+//         std::begin(c), sentinel<std::int16_t>{50}, std::begin(d), pred);
+//     auto post_result = std::count_if(std::begin(c), std::end(c), pred);
+
+//     HPX_TEST(pre_result == 2 && post_result == 1);
+// }
```
Comment (hkaiser) [inline libs/parallelism/algorithms/tests/unit/container_algorithms/remove_copy_if_range.cpp:46]: Can this comment be removed?

File: libs/parallelism/algorithms/include/hpx/parallel/algorithms/remove_copy.hpp (hunk 1 above)
Comment (gonidelis) [reply libs/parallelism/algorithms/include/hpx/parallel/algorithms/remove_copy.hpp:40]: Do we always declare our sequential algorithms as `constexpr` (for optimization reasons maybe)?

File: libs/parallelism/algorithms/tests/unit/container_algorithms/remove_copy_if_range.cpp (hunk 4 above)
Comment (gonidelis) [reply libs/parallelism/algorithms/tests/unit/container_algorithms/remove_copy_if_range.cpp:46]: This is a valid test that needs the recent changes on our custom sentinel (iter_sent.hpp) in order to run successfuly. I could either rebase on the unmerged branch (maybe?) or wait for #5125 to get merged.

File: libs/parallelism/algorithms/include/hpx/parallel/container_algorithms/remove_copy.hpp [hunk 5]
```diff
@@ -208,3 +208,236 @@ namespace hpx { namespace parallel { inline namespace v1 {
...
+                >
+            )>
+        // clang-format on
+        friend typename parallel::util::detail::algorithm_result<ExPolicy,
+            remove_copy_if_result<
+                typename hpx::traits::range_iterator<Rng>::type, O>>::type
+        tag_invoke(hpx::ranges::remove_copy_if_t, ExPolicy&& policy, Rng&& rng,
+            O dest, Pred&& pred, Proj&& proj = Proj())
+        {
+            typedef hpx::is_sequenced_execution_policy<ExPolicy> is_seq;
```
Comment (hkaiser) [inline libs/parallelism/algorithms/include/hpx/parallel/container_algorithms/remove_copy.hpp:332]: Could you please add `static_assert`s checking the iterator categories?

**Review by hkaiser — DISMISSED** [review]
Otherwise, LGTM, thanks!

File: libs/parallelism/algorithms/tests/unit/container_algorithms/remove_copy_if_range.cpp (hunk 4 above)
Comment (hkaiser) [reply libs/parallelism/algorithms/tests/unit/container_algorithms/remove_copy_if_range.cpp]: @gonidelis #5125 has been merged now, could you update this, please?

File: libs/parallelism/algorithms/include/hpx/parallel/algorithms/remove_copy.hpp (hunk 1 above)
Comment (hkaiser) [reply libs/parallelism/algorithms/include/hpx/parallel/algorithms/remove_copy.hpp]: Yes, if possible.

Comment (gonidelis) [reply libs/parallelism/algorithms/include/hpx/parallel/algorithms/remove_copy.hpp]: already have... sorry for not putting this as resolved ;)

File: libs/parallelism/algorithms/tests/unit/container_algorithms/remove_copy_if_range.cpp (hunk 4 above)
Comment (gonidelis) [reply libs/parallelism/algorithms/tests/unit/container_algorithms/remove_copy_if_range.cpp]: @hkaiser  already have... sorry for not putting this as resolved ;)

**Review by hkaiser — APPROVED** [review]
This is good to go, the remaining errors are unrelated.

**Pull request merged**